### ISSUE TITLE
socket activation (accept tcp listener file descriptor instead of binding to a tcp port)

### ISSUE THREAD

**Issue by tim-seoss** [static-web-server/static-web-server#36]
Firstly, thanks for writing static-web-server - it's something I've been wanting for a while!

Whilst looking at #35 I investigated how easy it would be to offer a socket activation model, as an alternative to binding to TCP port manually.

In this context it would mean that the parent process of static-web-server (e.g. inetd, systemd, or possibly Apple launchd) would open a TCP socket listener, and then pass the file descriptor for this listener to static-web-server.  In it's simplest form (and the model use by inetd), static-web-server would inherit the file descriptor for the TCP listener as it's standard input file descriptor (FD 0).  There is a Rust crate called `listenfd` designed to handle this situation (making the passed-in FD available as an `io::Result<Option<TcpListener>>`).

In the context of #35, this would allow:

- Static-web-server to bind to a port number below 1024 whilst not running as root (I know there are other methods to do this e.g. using firewalls and proxies, but this is probably the lowest overhead one).
- Restarting of the static-web-server without service interruption (e.g. for configuration change, or on SSL certificate renewal etc.) via https://www.freedesktop.org/software/systemd/man/systemd.socket.html#FlushPending=
- The use of various advanced socket binding options supported by systemd (e.g. `FreeBind`) without adding more code to static-web-server or warp to handle these.
- Further sandboxing of static-web-server from the host's network e.g. ref https://www.freedesktop.org/software/systemd/man/systemd.socket.html :

> All network sockets allocated through .socket units are allocated in the host's network namespace (see network_namespaces(7)). This does not mean however that the service activated by a configured socket unit has to be part of the host's network namespace as well. It is supported and even good practice to run services in their own network namespace (for example through PrivateNetwork=, see systemd.exec(5)), receiving only the sockets configured through socket-activation from the host's namespace. In such a set-up communication within the host's network namespace is only permitted through the activation sockets passed in while all sockets allocated from the service code itself will be associated with the service's own namespace, and thus possibly subject to a much more restrictive configuration.

Unfortunately using this model with Warp isn't as straightforward as I'd hoped, although it can be done via hyper as shown in this example:

https://github.com/rusty-crab/warp-api-starter-template/blob/833ca24994bd0a8e197332aee7df24eb687c84d1/src/main.rs#L172

(this uses the listenfd crate: https://docs.rs/listenfd/0.3.3/listenfd/index.html ).

Because static-web-server is already using a forked version of warp (and I'm not at all familiar with hyper and warp), I thought it best to just note this investigation at this point, rather than ploughing in and adding this feature to static-web-server.

Any thoughts?

**Comment by joseluisq** [comment]
> In the context of #35, this would allow:
> 
> * Static-web-server to bind to a port number below 1024 whilst not running as root (I know there are other methods to do this e.g. using firewalls and proxies, but this is probably the lowest overhead one).
> * Restarting of the static-web-server without service interruption (e.g. for configuration change, or on SSL certificate renewal etc.) via https://www.freedesktop.org/software/systemd/man/systemd.socket.html#FlushPending=
> * The use of various advanced socket binding options supported by systemd (e.g. `FreeBind`) without adding more code to static-web-server or warp to handle these.
> * Further sandboxing of static-web-server from the host's network e.g. ref https://www.freedesktop.org/software/systemd/man/systemd.socket.html :
> 
> > All network sockets allocated through .socket units are allocated in the host's network namespace (see network_namespaces(7)). This does not mean however that the service activated by a configured socket unit has to be part of the host's network namespace as well. It is supported and even good practice to run services in their own network namespace (for example through PrivateNetwork=, see systemd.exec(5)), receiving only the sockets configured through socket-activation from the host's namespace. In such a set-up communication within the host's network namespace is only permitted through the activation sockets passed in while all sockets allocated from the service code itself will be associated with the service's own namespace, and thus possibly subject to a much more restrictive configuration.

As being me not a systemd confident dev, I see socket activation feature in general as an interesting alternative.
I also saw the advantages described such as significant improvements specially for running non-on-demand or non-Internet services (please correct me if I'm wrong ?) along with the ability of rootless by default.

So why not. 
Regarding that I would like to highlight one thing. That possible feature "could" also involve to have to provide some .deb and .rpm packages in near future for Linux users (systemd) and maybe for Mac users as well similarly (launchd). But of course this would come later. What do you think?

> Unfortunately using this model with Warp isn't as straightforward as I'd hoped, although it can be done via hyper as shown in this example:
> 
> https://github.com/rusty-crab/warp-api-starter-template/blob/833ca24994bd0a8e197332aee7df24eb687c84d1/src/main.rs#L172
> 
> (this uses the listenfd crate: https://docs.rs/listenfd/0.3.3/listenfd/index.html ).

I'm also no aware about if Warp provides that kind of support (or perhaps [could this kind of help us?](https://github.com/seanmonstar/warp/blob/master/examples/unix_socket.rs#L13)) however since Warp is built upon Hyper I think would not be impossible.
BTW I saw that warp-api-starter is using systemfd + listenfd to provide debug mode with auto-reload feature. Is that correct? 

> Because static-web-server is already using a forked version of warp (and I'm not at all familiar with hyper and warp), I thought it best to just note this investigation at this point, rather than ploughing in and adding this feature to static-web-server.

Yes, static-web-server is using a synced fork of latest Warp but basically due to few features needed by this server that I have introduced to Warp's built-in filters only, concretely on headers, compression and dir ones (you can see them under [Jan 20, 2021 commits](https://github.com/joseluisq/warp/commits/master) if you are interested).
Honestly I have no plans to maintain those changes for long-time as a fork but since the server v2 is under beta so it means we can still continue hacking and of course it's not a blocking condition for future additions/improvements to this server.

--

Summing up, we can try to play around with listenfd/other along with Warp in order to see how could be addressed, from my side I will try to give a test. So feel free to let me know if there is progress about it. Thanks.

**Comment by tim-seoss** [comment]
I think it would be quite straightforward to implement the same design as the one referred to in the warp-api-starter-template above.  I think socket activation would be a useful feature for `static-web-server` even if it's not shipped as an OS package, but socket activation would probably be the logical way to package `static-web-server` (to allow maximum security in the default setup).  I haven't created a Debian package for a Rust binary before, but I've written some Rust, and done some Debian (and rpm) packaging, so I'd be happy to look at putting together a Debian package if it's implemented.

**Comment by joseluisq** [comment]
Great! in that case let's try it out. Just pinging me if you need review.

**Comment by joseluisq** [comment]
Just to let it know I was re-thinking the idea of Warp adoption for `static-web-server` project so [here](https://github.com/joseluisq/static-web-server/issues/37) the result.

**Comment by tim-seoss** [comment]
OK thanks.  I think it's probably better to wait on #37 (whatever eventual outcome), and then decide how to implement this feature afterwards.

**Comment by joseluisq** [comment]
#38 is finally merged so it leaves the way clear to address this feature issue.

**Comment by tim-seoss** [comment]
How would you like to divide the work on this?  I'm happy to have a go if you like?  It may be a simple revision...

**Comment by joseluisq** [comment]
Feel free to start when you want from whatever place that you feel is suitable.
I think you also want to have a look at [server.rs](https://github.com/joseluisq/static-web-server/blob/master/src/server.rs).

From my side I will support with review and changes if they are needed.
Since I'm focusing on other pending features like cors and directory listing. BTW those doesn't interfere with this feature.

**Comment by tim-seoss** [comment]
OK, sounds good.  Didn't want to duplicate effort if you were already looking at this.
I have a maximally hacky (minimal effort) proof of concept built (and serving successfully), so I'll try and find some time to finish it up and create a PR.
